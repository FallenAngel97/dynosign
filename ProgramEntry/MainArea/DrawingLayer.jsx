import React from 'react';
import { connect } from 'react-redux';
import { addLine, changeLayer, changeActiveLayer } from '../actions';
import PropTypes from 'prop-types';
import { Shape } from './Shape';
import { dragDroppedPicture, draw } from './drawingHelpers';

/**
 * @module DrawingLayer
 */

export class DrawingLayer extends React.Component {

  prevX = 0;
  currX = 0;
  prevX = 0;
  prevY = 0;
  dotFlag = false;
  flag = false;
  ctrlPressed = false;
  dotWasDrawn = false;

  constructor (props) {
    super(props);
    this.shapes = this.props.layer.shapes;
  }
  /**
   * Draws in mousemove process
   */
  draw = () => {
    const { changeColor, changeMouseType } = this.props;

    draw(this.canvas, changeColor.color, changeMouseType.mouseType, this.ctrlPressed, {
      prevX: this.prevX, prevY: this.prevY, currX: this.currX, currY: this.currY
    })
  }
  /**
   * Moves shapes when moveTool is selected in LeftBar
   * @param {string} res - Type of event
   * @param {object} e - mouse object
   */
  moveShape = (res, e) => {
    if (res === 'down') {
      this.prevX = this.currX;
      this.prevY = this.currY;

      this.currX = e.clientX - this.canvas.offsetLeft;
      this.currY = e.clientY - this.canvas.offsetTop;
      this.flag = true;
    }
    if (res === 'up') {
      this.flag = false;
    }
    if (res === 'move') {
      if (this.flag) {
        this.prevX = this.currX;
        this.prevY = this.currY;
        this.currX = e.clientX - this.canvas.offsetLeft;
        this.currY = e.clientY - this.canvas.offsetTop;
        const k = this.shapes.length;
        for (let i = k - 1; i >= 0; i--) {
          if (this.shapes[i].contains(this.currX, this.currY)) {
            this.shapes[i].x = this.shapes[i].x + (this.currX - this.prevX);
            this.shapes[i].y = this.shapes[i].y + (this.currY - this.prevY);
          }
        }
        this.draw();
      }
    }
  }
  pasteToCanvas = (e) => {
    console.log('paste');
    var items = e.clipboardData.items;
    console.log(items)
  }

  /**
   * Obtains coordinates of mouse over canvas and type of interaction of mouse
   * Hold ctrl to draw straight line
   * Hold shift and click points to connect points with straight line
   * @param {string} res - Type of event
   * @param {object} e - mouse object
   */
  findxy = (res, e) => {
    if (this.props.changeMouseType.mouseType === 'default') {
      this.moveShape(res, e);
      return;
    }
    if (!/draw|circle|rectangle/.test(this.props.changeMouseType.mouseType)) return;
    this.ctrlPressed = false;
    if (e.ctrlKey) {
      this.ctrlPressed = true;
    }
    if (res === 'down') {
      this.prevX = this.currX;
      this.prevY = this.currY;

      this.currX = (e.clientX - this.canvas.offsetLeft) * window.devicePixelRatio;
      this.currY = (e.clientY - this.canvas.offsetTop) * window.devicePixelRatio;

      if (/rectangle|circle/.test(this.props.changeMouseType.mouseType) || this.ctrlPressed) {
        this.prevX = this.currX;
        this.prevY = this.currY;
      }

      this.flag = true;
      if (this.props.changeMouseType.mouseType === 'draw') {
        this.dotWasDrawn = true;
        this.dotFlag = true;
      }
      if (this.dotFlag) {
        var ctx = this.canvas.getContext('2d');
        ctx.beginPath();
        const color = this.props.changeColor.color;
        if (e.shiftKey) {
          ctx.strokeStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
          ctx.moveTo(this.prevX, this.prevY);
          ctx.lineTo(this.currX, this.currY);
          ctx.stroke();
        } else {
          ctx.fillStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
          ctx.fillRect(this.currX, this.currY, 2, 2);
          ctx.closePath();
        }
        this.dotFlag = false;
      }
    }
    if (res === 'up' || res === 'out') {
      this.flag = false;
    }

    if (res === 'up') {
      const { layerNumber } = this.props.changeActiveLayer;
      this.props.addLine(this.canvas, layerNumber);
      this.props._changeActiveLayer(this.props.layersCRUD[layerNumber], layerNumber);
      const ctx = this.canvas.getContext('2d');
      const diffX = this.currX - this.prevX;
      const diffY = this.currY - this.prevY;
      if (this.props.changeMouseType.mouseType === 'rectangle') {
        this.shapes.push(new Shape(this.currX - diffX, this.currY - diffY, diffX, diffY, ctx.fillStyle));
      } else if (this.props.changeMouseType.mouseType === 'circle') {
        var circle = new Shape(this.currX - diffX, this.currY - diffY, Math.abs(diffX), diffY, ctx.fillStyle);
        circle.setShape('circle');
        this.shapes.push(circle);
      }
    }

    if (res === 'move') {
      if (this.flag) {
        this.dotWasDrawn = false;
        if (this.props.changeMouseType.mouseType === 'draw') {
          if (!this.ctrlPressed)
            this.prevX = this.currX;
          this.prevY = this.currY;
        }
        this.currX = (e.clientX - this.canvas.offsetLeft) * window.devicePixelRatio;
        this.currY = (e.clientY - this.canvas.offsetTop) * window.devicePixelRatio;
        this.draw();
      }
    }
  }

  /**
   * Allows to drop the image from the folder
   * @param {MouseEvent} e - necessary to disable the default drop behaviour
   */

  onDrop = (e) => {
    e.preventDefault();
    const droppedFileName = dragDroppedPicture(this.canvas, e);
    let { layer } = this.props.changeActiveLayer;
    layer.name = droppedFileName;
    this.props.changeLayer(layer, this.props.changeActiveLayer.layerNumber)
    console.log(droppedFileName);
    return false;
  }

  render () {
    const { width, height, layer } = this.props;

    return (
      <canvas className='canvaslayer'
        width={width * window.devicePixelRatio}
        height={height * window.devicePixelRatio}
        style={{ width, height, opacity: parseInt(this.props.layer.opacity) / 100 }}
        onMouseMove={(e) => this.findxy('move', e)}
        onMouseDown={(e) => this.findxy('down', e)}
        onDragOver={(e) => { e.preventDefault(); return false } }
        onDragLeave={() => false }
        onDragStart={(ev) => ev.dataTransfer.setData('Text', ev.target.id) }
        onDragEnd={() => false }
        onDrop={this.onDrop}
        onMouseUp={(e) => this.findxy('up', e)}
        onMouseOut={(e) => this.findxy('out', e)}
        onPaste={(e) => this.pasteToCanvas(e)}
        ref={(canvas) => { this.canvas = canvas }} />
    )
  }
}

DrawingLayer.propTypes = {
  changeColor: PropTypes.object,
  changeMouseType: PropTypes.object,
  changeLayer: PropTypes.func,
  height: PropTypes.number,
  width: PropTypes.number,
  addLine: PropTypes.func,
  changeActiveLayer: PropTypes.object,
  layer: PropTypes.object
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    addLine: (elem, layerNumber) => dispatch(addLine(elem, layerNumber)),
    changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
    _changeActiveLayer :(layer, layerNumber) => dispatch(changeActiveLayer(layer, layerNumber))
  }
}

export default connect(mapStateToProps, mapDispatchToProps, null, { forwardRef: true })(DrawingLayer);
