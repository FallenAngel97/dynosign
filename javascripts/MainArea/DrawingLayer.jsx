import React from "react";
import {connect} from "react-redux";
import {addLine} from "../actions";

export class DrawingLayer extends React.Component {
    constructor(props) {
        super(props);
        this.prevX = 0;
        this.currX = 0;
        this.prevX = 0;
        this.prevY = 0;
        this.dot_flag = false;
        this.flag = false;
        this.findxy = this.findxy.bind(this);
        this.draw = this.draw.bind(this);
    }
    draw() {
        var ctx = this.canvas.getContext('2d');
        ctx.beginPath();
        if(this.props.changeMouseType.mouseType == 'draw') {
            ctx.restore();
            ctx.moveTo(this.prevX, this.prevY);
            ctx.lineTo(this.currX, this.currY);
            var x = "black",
            y = 2;
            ctx.strokeStyle = x;
            ctx.lineWidth = y;
            ctx.stroke();
            ctx.closePath();
        }
        else {
            ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
            ctx.setLineDash([6])
            const diffX = this.currX - this.prevX;
            const diffY = this.currY - this.prevY;
            ctx.strokeRect(this.currX - diffX, this.currY - diffY, diffX,diffY);
        }
    }
    findxy(res, e) {
        if (res == 'down') {
            this.prevX = this.currX;
            this.prevY = this.currY;

            this.currX = e.clientX - this.canvas.offsetLeft;
            this.currY = e.clientY - this.canvas.offsetTop;
            if(this.props.changeMouseType.mouseType == 'select') {
                this.prevX = this.currX;
                this.prevY = this.currY;
            }
            this.flag = true;
            this.dot_flag = true;
            if (this.dot_flag) {
                var ctx = this.canvas.getContext('2d');
                ctx.beginPath();
                ctx.fillStyle = 'black';
                ctx.fillRect(this.currX, this.currY, 2, 2);
                ctx.closePath();
                this.dot_flag = false;
            }
        }
        if (res == 'up' || res == "out") {
            this.flag = false;
        }
        
        if(res=="up"){
            console.log()
            if(this.props.changeMouseType.mouseType == 'draw')
                this.props.addLine(this.canvas, this.props.changeActiveLayer.layerNumber);
            else {

            }
        }
        
        if (res == 'move') {
            if (this.flag) {
                if(this.props.changeMouseType.mouseType == 'draw') {
                    this.prevX = this.currX;
                    this.prevY = this.currY;
                }
                this.currX = e.clientX - this.canvas.offsetLeft;
                this.currY = e.clientY - this.canvas.offsetTop;
                this.draw();
            }
        }
    }
    render() {
        return (
            <canvas className='canvaslayer'
                    width={this.props.width}
                    height={this.props.height}
                    style={{width: this.props.width, height: this.props.height}}
                    onMouseMove={(e)=> this.findxy('move', e)}
                    onMouseDown={(e)=> this.findxy('down', e)}
                    onMouseUp={(e)=> this.findxy('up', e)}
                    onMouseOut={(e)=> this.findxy('out', e)}
                    ref={(canvas)=>this.canvas = canvas} />
        )
    }
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
    return {
        addLine: (elem, layerNumber) => dispatch(addLine(elem, layerNumber)),
    }
}

export default connect(mapStateToProps, mapDispatchToProps,null, {forwardRef: true})(DrawingLayer);