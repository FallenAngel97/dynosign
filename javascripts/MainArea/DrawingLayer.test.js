import React from "react";
import {mapDispatchToProps, DrawingLayer} from "./DrawingLayer";
import { configure, shallow, mount} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() })
HTMLCanvasElement.prototype.toDataURL = () => {
    return 'canvas-data-url';
}

describe("DrawingLayer test set", () => {
    test("DrawingLayer can add line", () => {
        const dispatch = jest.fn();
        const canvas = document.createElement('canvas');
        const layerNumber = 2;
        const expectedAction = {
            type: 'ADD_LINE', lineData: canvas.toDataURL(), layerNumber
        }
        mapDispatchToProps(dispatch).addLine(canvas, layerNumber);
        expect(dispatch.mock.calls[0][0]).toEqual(expectedAction);
    });
    test("DrawingLayer renders properly", () => {
        const layer = mount(<DrawingLayer />);
        expect(layer).toMatchSnapshot();
        expect(layer.instance().canvas).toBeTruthy();
    });
});