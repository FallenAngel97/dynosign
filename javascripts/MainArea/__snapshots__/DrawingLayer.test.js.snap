// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawingLayer test set DrawingLayer renders properly 1`] = `ReactWrapper {}`;
